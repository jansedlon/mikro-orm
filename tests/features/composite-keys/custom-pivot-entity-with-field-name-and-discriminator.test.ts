import {
  Collection,
  Entity,
  EntityCaseNamingStrategy,
  ManyToMany,
  ManyToOne,
  MikroORM,
  PrimaryKey,
  Property,
} from '@mikro-orm/core';

@Entity({ abstract: true })
export abstract class AbstractEntity {

  @PrimaryKey()
  public id!: number;

}

@Entity({
  discriminatorColumn: 'role',
  discriminatorMap: {
    CREATOR: 'Creator',
  },
})
class User extends AbstractEntity {

  @Property({ default: 'CREATOR' })
  public role: 'CREATOR' = 'CREATOR' as const;

}

@Entity()
class Creator extends User {

  // eslint-disable-next-line @typescript-eslint/no-use-before-define
  @ManyToMany({ entity: () => Task, pivotEntity: () => CreatorsOnTasks })
  public tasks = new Collection<Task>(this);

}

@Entity({ tableName: 'Task' })
class Task extends AbstractEntity {

  @ManyToMany(() => Creator, c => c.tasks)
  public creators = new Collection<Creator>(this);

}

@Entity({ tableName: 'CreatorsOnTasks' })
class CreatorsOnTasks {

  @ManyToOne({ primary: true, fieldName: 'creatorId', entity: () => Creator })
  public creator!: Creator;

  @ManyToMany({ primary: true, fieldName: 'taskId', entity: () => Task })
  public task!: Task;

}

describe('mikroOrm', () => {
  let orm: MikroORM;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [User, Creator, CreatorsOnTasks, Task],
      dbName: ':memory:',
      type: 'sqlite',
      namingStrategy: EntityCaseNamingStrategy,
    });
    await orm.schema.createSchema();
  });

  afterAll(async () => await orm.close(true));

  beforeEach(() => orm.schema.clearDatabase());

  test('schema', async () => {
    const sql = await orm.schema.getCreateSchemaSQL();

    expect(sql).toMatchSnapshot();
  });

  async function createEntities() {
    const task = new Task();
    const creator = new Creator();

    task.creators.add(creator);

    await orm.em.fork().persistAndFlush(task);

    return { task };
  }

  test('should insert', async () => {
    await expect(createEntities()).resolves.not.toThrow();
  });

  test('shouldnt findOne and populate m:n relation', async () => {
    const { task } = await createEntities();

    const taskRepository = orm.em.fork().getRepository(Task);
    await expect(
      taskRepository.findOne({ id: task.id }, { populate: ['creators'] }),
    ).resolves.not.toThrow();
  });

});
