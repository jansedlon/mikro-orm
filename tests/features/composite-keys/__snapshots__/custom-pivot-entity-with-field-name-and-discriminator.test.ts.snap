// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mikroOrm schema 1`] = `
"pragma foreign_keys = off;

create table \`Task\` (\`id\` integer not null primary key autoincrement);

create table \`User\` (\`id\` integer not null primary key autoincrement, \`role\` text not null default 'CREATOR');
create index \`User_role_index\` on \`User\` (\`role\`);

create table \`CreatorsOnTasks\` (\`creatorId\` integer not null, \`taskId\` integer not null, constraint \`CreatorsOnTasks_creatorId_foreign\` foreign key(\`creatorId\`) references \`User\`(\`id\`) on update cascade, primary key (\`creatorId\`, \`taskId\`));
create index \`CreatorsOnTasks_creatorId_index\` on \`CreatorsOnTasks\` (\`creatorId\`);

create table \`CreatorsOnTasks_task\` (\`creatorsOnTasks_creatorId\` integer not null, \`creatorsOnTasks_taskId\` integer not null, \`task\` integer not null, constraint \`CreatorsOnTasks_task_creatorsOnTasks_creatorId_creatorsOnTasks_taskId_foreign\` foreign key(\`creatorsOnTasks_creatorId\`, \`creatorsOnTasks_taskId\`) references \`CreatorsOnTasks\`(\`creatorId\`, \`taskId\`) on delete cascade on update cascade, constraint \`CreatorsOnTasks_task_task_foreign\` foreign key(\`task\`) references \`Task\`(\`id\`) on delete cascade on update cascade, primary key (\`creatorsOnTasks_creatorId\`, \`creatorsOnTasks_taskId\`, \`task\`));
create index \`CreatorsOnTasks_task_task_index\` on \`CreatorsOnTasks_task\` (\`task\`);
create index \`CreatorsOnTasks_task_creatorsOnTasks_creatorId_creatorsOnTasks_taskId_index\` on \`CreatorsOnTasks_task\` (\`creatorsOnTasks_creatorId\`, \`creatorsOnTasks_taskId\`);

pragma foreign_keys = on;
"
`;
